sympy/solvers/ode.py

```diff
@@ -5720,6 +5724,110 @@ def checkinfsol(eq, infinitesimals, func=None, order=None):
+        r = pdsolve(rpde, func=f(x, y)).rhs
+        s = pdsolve(rpde - 1, func=f(x, y)).rhs
+        newcoord = [_lie_group_remove(coord) for coord in [r, s]]
+        r = Dummy("r")
+        s = Dummy("s")
+        C1 = Symbol("C1")
+        rcoord = newcoord[0]
+        scoord = newcoord[-1]
+        try:
+            sol = solve([r - rcoord, s - scoord], x, y, dict=True)
+            if sol == []:
+                continue
+        except NotImplementedError:
+            continue
+        else:
+            sol = sol[0]
+            xsub = sol[x]
+            ysub = sol[y]
+            num = simplify(scoord.diff(x) + scoord.diff(y)*h)
+            denom = simplify(rcoord.diff(x) + rcoord.diff(y)*h)
+            if num and denom:
+                diffeq = simplify((num/denom).subs([(x, xsub), (y, ysub)]))
+                sep = separatevars(diffeq, symbols=[r, s], dict=True)
+                if sep:
+                    # Trying to separate, r and s coordinates
+                    deq = integrate((1/sep[s]), s) + C1 - integrate(sep['coeff']*sep[r], r)
+                    # Substituting and reverting back to original coordinates
+                    deq = deq.subs([(r, rcoord), (s, scoord)])
+                    try:
+                        sdeq = solve(deq, y)
+                    except NotImplementedError:
+                        tempsol.append(deq)
+                    else:
+                        return [Eq(f(x), sol) for sol in sdeq]
+
+
+            elif denom: # (ds/dr) is zero which means s is constant
+                return [Eq(f(x), solve(scoord - C1, y)[0])]
+
+            elif num: # (dr/ds) is zero which means r is constant
+                return [Eq(f(x), solve(rcoord - C1, y)[0])]
+
+    # If nothing works, return solution as it is, without solving for y
+    if tempsol:
+        return [Eq(sol.subs(y, f(x)), 0) for sol in tempsol]
+
+def _ode_lie_group( s, func, order, match):
+    heuristics = lie_heuristics
+    inf = {}
+    f = func.func
+    x = func.args[0]
+    df = func.diff(x)
+    xi = Function("xi")
+    eta = Function("eta")
+    xis = match['xi']
+    etas = match['eta']
+    y = match.pop('y', None)
+    if y:
+        h = -simplify(match[match['d']]/match[match['e']])
+        y = y
+    else:
+        y = Dummy("y")
+        h = s.subs(func, y)
+
+    if xis is not None and etas is not None:
+        inf = [{xi(x, f(x)): S(xis), eta(x, f(x)): S(etas)}]
+
+        if checkinfsol(Eq(df, s), inf, func=f(x), order=1)[0][0]:
+            heuristics = ["user_defined"] + list(heuristics)
+
+    match = {'h': h, 'y': y}
+
+    # This is done so that if:
+    # a] solve raises a NotImplementedError.
+    # b] any heuristic raises a ValueError
+    # another heuristic can be used.
+    for heuristic in heuristics:
+        try:
+            return _ode_lie_group_try_heuristic(Eq(df, s), heuristic, func, match, inf)
+        except ValueError:
+            continue
+    return []
```

Comment:
The convention for `dsolve` is to raise `NotImplementedError` when an ODE can't be solved:
```julia
In [8]: dsolve(f(x).diff(x, 2)-exp(f(x)))                                                                                         
---------------------------------------------------------------------------
NotImplementedError
```
We shouldn't return an empty list unless we *know* that there are no solutions. That might happen as a result of the initial conditions being inconsistent for example. I don't think we ever know from the Lie group methods that there are no solutions because there could always be more heuristics to try (I might be wrong about that though).

Reply:
>I don't think we ever know from the Lie group methods that there are no solutions because there could always be more heuristics to try (I might be wrong about that though).

Yes, we can't say ever that solution doesn't exist if we are not able to solve for the predefined `heuristic`.

Reply:
Here I am returning an empty list because I am trying to find solutions for every element in  `sol=solve(eq, df)` so we can't raise `NotImplementedError` without checking for every term in `sol`.
But in the end, when we are not getting any solutions the I am raising `NotImplementedError`.
And I don't know why you have given that example and the given example is not solvable by any method.

```
In [7]: classify_ode(f(x).diff(x, 2)-exp(f(x)))
Out[7]: ()
```